File: backend-kotlin/src/main/kotlin/com/bffgin/backend/auth/TokenVerifier.kt
package com.bffgin.backend.auth

/**
 * suspend funにしているのは、JwksVerifier実装がkid不一致時にJWKSをネットワーク越しに
 * 再取得する(Ktor Clientのsuspend呼び出し)必要があるため。HmacVerifierは実際には
 * 何も一時停止しない(jjwtの同期検証のみ)が、インターフェースは統一している
 */
fun interface TokenVerifier {
    suspend fun verify(token: String): Claims
}

File: backend-python/app/auth/token_verifier.py
from __future__ import annotations

from typing import Protocol

from app.auth.claims import Claims


class TokenVerifier(Protocol):
    """JwksVerifier実装がkid不一致時にJWKSをネットワーク越しに再取得する(httpxのasync呼び出し)
    必要があるため、非同期メソッドにしている。HmacVerifierは実際には何もawaitしない
    (PyJWTの同期検証のみ)が、インターフェースは統一している
    """

    async def verify(self, token: str) -> Claims: ...
